src/rgw/driver/d4n/rgw_sal_d4n.cc

```diff
@@ -1816,48 +1812,85 @@ int D4NFilterObject::D4NFilterDeleteOp::delete_obj(const DoutPrefixProvider* dpp
-    ldpp_dout(dpp, 0) << "D4NFilterObject::D4NFilterDeleteOp::" << __func__ << "(): calling next delete_obj" << dendl;
-    return next->delete_obj(dpp, y, flags);
+    ldpp_dout(dpp, 10) << "D4NFilterObject::" << __func__ << "(): head object not found; calling next->delete_obj" << dendl;
+    next->params = params;
+    int ret = next->delete_obj(dpp, y, flags);
+    ret = next->delete_obj(dpp, y, flags);
+    result = next->result;
+    return ret;
   } else {
     int ret = -1;
     bool objDirty = false;
     auto blockDir = source->driver->get_block_dir(); 
     std::string version, policy_prefix;
+    bool nullDelete = false;
 
     if (!source->get_bucket()->versioned()) {
       version = source->get_object_version();
     } else if (source->get_bucket()->versioned() && !source->have_instance()) {
-      rgw::d4n::CacheBlock deleteBlock;
-      block.prevVersion = std::pair<std::string, bool>(block.version, block.deleteMarker);
-      block.deleteMarker = true;
+      auto tempBlock = block;
+
+      if (!source->get_bucket()->versioning_enabled()) { // if versioning is suspended
+        if (block.version == source->get_object_version()) {
+	  version = source->get_object_version();
+	  block.prevVersion = std::pair<std::string, bool>(block.version, block.deleteMarker);
+	  block.deleteMarker = true;
+	  block.version = "null";
 
-      if (source->get_bucket()->versioned() && !source->get_bucket()->versioning_enabled()) { // if versioning is suspended
-        block.version = "null"; 
+	  nullDelete = true; // indicates a simple delete request was made and the object already has a null versionId;
+			     // therefore, delete the null version and add a null delete marker on top
+        } else {
+	  tempBlock.cacheObj.objName = "_:null_" + block.cacheObj.objName;
+	  ret = blockDir->get(dpp, &tempBlock, y); // check if null delete marker block already exists; if so, delete
+	  if (ret == 0 && tempBlock.deleteMarker) {
+	    version = "null";
+	  } else if (ret == -ENOENT) { // if there is no null delete marker block to delete, add one
+	    block.prevVersion = std::pair<std::string, bool>(block.version, block.deleteMarker);
+	    block.deleteMarker = true;
+	    block.version = "null";
+	  } else {
+	    ldpp_dout(dpp, 0) << "Failed to retrieve null head object in block directory, ret=" << ret << dendl;
+	    return ret;
+	  }
+        }
       } else {
-	// create a delete marker
-	enum { OBJ_INSTANCE_LEN = 32 };
-	char buf[OBJ_INSTANCE_LEN + 1];
-	gen_rand_alphanumeric_no_underscore(dpp->get_cct(), buf, OBJ_INSTANCE_LEN);
-        block.version = buf; // using gen_rand_alphanumeric_no_underscore for the time being
+	block.prevVersion = std::pair<std::string, bool>(block.version, block.deleteMarker);
+	block.deleteMarker = true;
+
+	ldpp_dout(dpp, 10) << "D4NFilterObject::" << __func__ << "(): retrieving delete marker; calling next->delete_obj" << dendl;
+	next->params = params;
+	ret = next->delete_obj(dpp, y, flags);
```

Comment:
next->delete_obj is being called without checking if the object is dirty or not.